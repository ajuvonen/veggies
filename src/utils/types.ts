import type {DateTime} from 'luxon';

export type Locale = 'en' | 'fi';

export enum Category {
  Fruit = 'Fruit',
  Vegetable = 'Vegetable',
  Leafy = 'Leafy',
  Root = 'Root',
  Bean = 'Bean',
  Grain = 'Grain',
  Mushroom = 'Mushroom',
}

export type Favorites = {
  [key in Category]: readonly [string, number][];
};

export type Week = {
  veggies: string[];
  readonly startDate: DateTime;
};

export type Challenge = {
  readonly startDate: DateTime;
  veggie: string | undefined;
};

export type Settings = {
  allergens: string[];
  locale: Locale;
  showChartAnimations: boolean;
  suggestionCount: number;
  summaryViewedDate: DateTime | null;
};

export type Listing = {
  readonly veggie: string;
  readonly category: Category;
};

export type TranslatedListing = Listing & {
  readonly translation: string;
  readonly synonyms: string[];
};

export enum AchievementLevel {
  NoAchievement,
  Bronze,
  Silver,
  Gold,
  Platinum,
}

export type Achievements = {
  allOnRed: AchievementLevel;
  botanicalBerries: AchievementLevel;
  challengeAccepted: AchievementLevel;
  committed: AchievementLevel;
  completionist: AchievementLevel;
  experimenterBean: AchievementLevel;
  experimenterFruit: AchievementLevel;
  experimenterGrain: AchievementLevel;
  experimenterLeafy: AchievementLevel;
  experimenterMushroom: AchievementLevel;
  experimenterRoot: AchievementLevel;
  experimenterVegetable: AchievementLevel;
  goNuts: AchievementLevel;
  hotStreak: AchievementLevel;
  lemons: AchievementLevel;
  thirtyVeggies: AchievementLevel;
  thousandsOdd: AchievementLevel;
  thousandsEven: AchievementLevel;
};

export type WeekData = {
  atMostVeggies: number;
  challenge: string | undefined;
  firstTimeVeggies: string[];
  firstWeek: boolean;
  hotStreak: number;
  mean: number;
  previousWeekCount: number;
  veggies: string[];
  weekNumber: string;
};

export type SummaryItem = {
  emoji: string;
  translationKey: string;
  translationParameters: (string | number)[];
};
