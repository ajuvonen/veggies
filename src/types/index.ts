export type WeekData = {
  veggies: string[];
  weekNumber: string;
};

export type SummaryItem = {
  emoji: string;
  translationKey: string;
  translationParameters: (string | number)[];
};
