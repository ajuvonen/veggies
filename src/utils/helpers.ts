import {useMemoize} from '@vueuse/core';
import type {ChartOptions, ChartType, Scale} from 'chart.js';
import type {Context} from 'chartjs-plugin-datalabels';
import {DateTime} from 'luxon';
import {mergeDeep, sample} from 'remeda';
import {
  ALL_VEGGIES,
  BEANS,
  CATEGORY_EMOJI,
  COLORS,
  FRUITS,
  GRAINS,
  LEAFIES,
  MUSHROOMS,
  ROOTS,
  VEGETABLES,
} from '@/utils/constants';
import {AchievementLevel, Category} from '@/utils/types';

export const getCategoryForVeggie = useMemoize((veggie: string) => {
  if (FRUITS.includes(veggie)) {
    return Category.Fruit;
  } else if (VEGETABLES.includes(veggie)) {
    return Category.Vegetable;
  } else if (LEAFIES.includes(veggie)) {
    return Category.Leafy;
  } else if (ROOTS.includes(veggie)) {
    return Category.Root;
  } else if (BEANS.includes(veggie)) {
    return Category.Bean;
  } else if (GRAINS.includes(veggie)) {
    return Category.Grain;
  } else if (MUSHROOMS.includes(veggie)) {
    return Category.Mushroom;
  }
  return undefined;
});

export const getChartOptions = <T extends ChartType>(
  grids: boolean,
  stacked: boolean,
  withIcons: boolean,
  showChartAnimations: boolean,
  overrides: Partial<ChartOptions<T>> = {},
) =>
  mergeDeep(
    {
      animation: !showChartAnimations ? false : undefined,
      responsive: true,
      maintainAspectRatio: !grids,
      normalized: true,
      layout: {
        padding: 0,
      },
      scales: grids
        ? {
            y: {
              beginAtZero: true,
              ticks: {
                precision: 0,
                color: COLORS.offWhite,
              },
              stacked,
            },
            y1: {
              position: 'right',
              beginAtZero: true,
              ticks: {
                precision: 0,
                color: COLORS.offWhite,
              },
              stacked,
              afterBuildTicks: (axis: Scale) => {
                axis.ticks = [...axis.chart.scales.y.ticks];
                axis.min = axis.chart.scales.y.min;
                axis.max = axis.chart.scales.y.max;
              },
            },
            x: {
              beginAtZero: true,
              ticks: {
                precision: 0,
                color: COLORS.offWhite,
              },
              stacked,
            },
          }
        : undefined,
      plugins: {
        title: {
          display: false,
        },
        legend: {
          display: false,
        },
        datalabels: {
          ...(withIcons
            ? {
                anchor: 'center',
                align: 'center',
                font: {
                  size: 25,
                },
                textShadowColor: '#fff',
                textShadowBlur: 5,
                // eslint-disable-next-line @typescript-eslint/no-explicit-any
                formatter: (value: any, {dataset: {label}}: Context) =>
                  value ? CATEGORY_EMOJI[label as Category] : '',
              }
            : {
                display: false,
              }),
        },
        tooltip: {
          animation: showChartAnimations,
          padding: 8,
          titleFont: {
            size: 14,
            weight: 'normal',
          },
          bodyFont: {
            size: 14,
          },
          footerFont: {
            size: 14,
            weight: 'normal',
          },
          displayColors: false,
          backgroundColor: COLORS.darkBlue,
          bodyColor: COLORS.offWhite,
          titleColor: COLORS.offWhite,
        },
      },
    },
    overrides,
  ) as ChartOptions<T>;

export const getRandomVeggie = () => ALL_VEGGIES[Math.floor(Math.random() * ALL_VEGGIES.length)];

// eslint-disable-next-line @typescript-eslint/no-explicit-any
export const dateParser = (key: string, value: any) => {
  if (key === 'startDate' && value) {
    return DateTime.fromISO(value.split('T')[0]);
  }
  return value;
};

// eslint-disable-next-line @typescript-eslint/no-explicit-any
export const dateReplacer = (key: string, value: any) => {
  if (key === 'startDate' && value) {
    return value.split('T')[0];
  }
  return value;
};

const veggieEmojis: readonly string[] = [
  '🥝',
  '🥥',
  '🍇',
  '🍈',
  '🍉',
  '🍊',
  '🍋',
  '🍌',
  '🍍',
  '🥭',
  '🍎',
  '🍏',
  '🍐',
  '🍑',
  '🍒',
  '🍓',
  '🫐',
  '🍅',
  '🍆',
  '🌽',
  '🌶️',
  '🫑',
  '🥑',
  '🥒',
  '🥬',
  '🥦',
  '🥔',
  '🧄',
  '🧅',
  '🥕',
  '🫛',
  '🥜',
  '🫘',
  '🌿',
  '🍄‍🟫',
  '🌾',
  '🌰',
];
export const getRandomEmojis = (amount: number = 1) => sample(veggieEmojis, amount);

export const achievementLevelHelper = (levels: [number, AchievementLevel][], value: number) =>
  levels.find(([threshold]) => value >= threshold)?.[1] ?? AchievementLevel.NoAchievement;
