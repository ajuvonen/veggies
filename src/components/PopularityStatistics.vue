<script lang="ts" setup>
import {storeToRefs} from 'pinia';
import {useActivityStore} from '@/stores/activityStore';
import ContentElement from '@/components/ContentElement.vue';

const {categoryFavorites} = storeToRefs(useActivityStore());
</script>
<template>
  <div class="popularity__container">
    <ContentElement
      v-for="[category, favorites] in categoryFavorites"
      :key="category"
      :title="$t(`popularity.${category}`)"
    >
      <ul class="flex-container flex-col align-start">
        <li
          v-for="(emoji, index) in ['ðŸ¥‡', 'ðŸ¥ˆ', 'ðŸ¥‰']"
          :key="emoji"
          class="popularity__entry"
          :data-test-id="`popularity-${category}-${index + 1}`"
        >
          <div class="popularity__entry-emoji" aria-hidden="true">{{ emoji }}</div>
          <div v-if="favorites[index]" class="popularity__entry-text">
            {{
              $t('popularity.entry', [$t(`veggies.${favorites[index][0]}`), favorites[index][1]])
            }}
          </div>
          <div v-else class="popularity__entry-text">{{ $t('popularity.noEntry') }}</div>
        </li>
      </ul>
    </ContentElement>
  </div>
</template>
<style lang="scss" scoped>
.popularity__container {
  @apply flex-container flex-col gap-4;
  @apply has-scroll;
}

.popularity__entry {
  @apply flex items-center;
}

.popularity__entry-emoji {
  @apply text-4xl;
  filter: drop-shadow(0 0 2px #fff);
}

.popularity__entry-text {
  @apply capitalize;
}

.popularity__honorable-mentions {
  @apply capitalize list-decimal list-inside;
}
</style>
