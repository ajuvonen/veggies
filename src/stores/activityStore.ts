import {computed, ref} from 'vue';
import {defineStore, storeToRefs} from 'pinia';
import {debounceFilter, useIntervalFn, useNow, useStorage} from '@vueuse/core';
import {DateTime} from 'luxon';
import {
  countBy,
  difference,
  entries,
  intersection,
  map,
  pipe,
  prop,
  sortBy,
  take,
  takeWhile,
} from 'remeda';
import {
  Category,
  type Favorites,
  type Challenge,
  type Week,
  type Achievements,
  AchievementLevel,
} from '@/utils/types';
import {
  achievementLevelHelper,
  dateParser,
  dateReplacer,
  getCategoryForVeggie,
  getRandomVeggie,
} from '@/utils/helpers';
import {useAppStateStore} from '@/stores/appStateStore';
import {
  BEANS,
  BOTANICAL_BERRIES,
  FRUITS,
  GRAINS,
  LEAFIES,
  MUSHROOMS,
  NUTS,
  RED_VEGGIES,
  ROOTS,
  VEGETABLES,
} from '@/utils/constants';

export const useActivityStore = defineStore('activity', () => {
  const {settings} = storeToRefs(useAppStateStore());
  const currentDate = ref(DateTime.now());
  useIntervalFn(() => {
    const now = DateTime.now();
    if (!currentDate.value.hasSame(now, 'day')) {
      currentDate.value = now;
    }
  }, 2000);

  // State refs
  const startDate = useStorage<DateTime | null>('veggies-start-date', null, localStorage, {
    mergeDefaults: true,
    serializer: {
      read: (v) => (v ? DateTime.fromISO(v.split('T')[0]) : null),
      write: (v) => v?.toISODate() ?? '',
    },
  });

  const weeks = useStorage<Week[]>('veggies-weeks', [], localStorage, {
    mergeDefaults: true,
    eventFilter: debounceFilter(2000),
    serializer: {
      read: (v) => (v ? JSON.parse(v, dateParser) : null),
      write: (v) => JSON.stringify(v, dateReplacer),
    },
  });

  const challenges = useStorage<Challenge[]>('veggies-challenges', [], localStorage, {
    mergeDefaults: true,
    serializer: {
      read: (v) => (v ? JSON.parse(v, dateParser) : null),
      write: (v) => JSON.stringify(v, dateReplacer),
    },
  });

  // Computed getters
  const currentWeekStart = computed(() => currentDate.value.startOf('week'));

  const getWeekStarts = computed(() => {
    if (!startDate.value) return [currentWeekStart.value];
    const totalWeeks = Math.ceil(currentDate.value.diff(startDate.value, 'week').weeks);
    return Array.from({length: totalWeeks}, (_, weekIndex) =>
      currentWeekStart.value.minus({weeks: weekIndex}),
    );
  });

  const hotStreak = computed(() => {
    let over30Weeks = takeWhile(
      getWeekStarts.value,
      (weekStart, weekIndex) => weekIndex === 0 || veggiesForWeek.value(weekStart).length >= 30,
    ).length;
    if (currentVeggies.value.length < 30) {
      over30Weeks--;
    }
    return over30Weeks;
  });

  const allVeggies = computed(() => weeks.value.flatMap(prop('veggies')));

  const uniqueVeggies = computed(() => [...new Set(allVeggies.value)]);

  const over30Veggies = computed(
    () => weeks.value.filter(({veggies}) => veggies.length >= 30).length,
  );

  const atMostVeggies = computed(() =>
    Math.max(...weeks.value.map(({veggies}) => veggies.length), 0),
  );

  const completedChallenges = computed(
    () =>
      challenges.value.filter(({startDate, veggie}) =>
        veggiesForWeek.value(startDate).includes(veggie),
      ).length,
  );

  const veggiesForWeek = computed(
    () => (weekStart: DateTime) =>
      weeks.value.find(({startDate}) => startDate.equals(weekStart))?.veggies ?? [],
  );

  const currentVeggies = computed({
    get: () => veggiesForWeek.value(currentWeekStart.value),
    set: (veggies: string[]) => setVeggiesForWeek(veggies),
  });

  const currentChallenge = computed(
    () => challenges.value.find(({startDate}) => startDate.equals(currentWeekStart.value))?.veggie,
  );

  const suggestions = computed(() =>
    pipe(
      allVeggies.value.filter((veggie) => !currentVeggies.value.includes(veggie)),
      countBy((veggie) => veggie),
      entries(),
      sortBy([prop(1), 'desc']),
      take(settings.value.suggestionCount),
      map(prop(0)),
    ),
  );

  const favorites = computed(() =>
    Object.values(Category).reduce(
      (acc, category) => ({
        ...acc,
        [category]: pipe(
          allVeggies.value.filter((veggie) => getCategoryForVeggie(veggie) === category),
          countBy((veggie) => veggie),
          entries(),
          sortBy([prop(1), 'desc']),
          take(6),
        ),
      }),
      {} as Favorites,
    ),
  );

  const weeklyAchievements = computed(() => (veggies: string[] = currentVeggies.value) => ({
    allOnRed: achievementLevelHelper(
      [[10, AchievementLevel.Gold]],
      intersection(RED_VEGGIES, veggies).length,
    ),
    botanicalBerries: achievementLevelHelper(
      [[15, AchievementLevel.Gold]],
      intersection(BOTANICAL_BERRIES, veggies).length,
    ),
    goNuts: achievementLevelHelper(
      [[5, AchievementLevel.Gold]],
      intersection(NUTS, veggies).length,
    ),
    thirtyVeggies: achievementLevelHelper(
      [
        [40, AchievementLevel.Platinum],
        [30, AchievementLevel.Gold],
      ],
      veggies.length,
    ),
  }));

  const achievements = computed<Achievements>(() => ({
    challengeAccepted: achievementLevelHelper(
      [
        [20, AchievementLevel.Gold],
        [10, AchievementLevel.Silver],
        [5, AchievementLevel.Bronze],
      ],
      completedChallenges.value,
    ),
    committed: achievementLevelHelper(
      [
        [52, AchievementLevel.Gold],
        [26, AchievementLevel.Silver],
        [12, AchievementLevel.Bronze],
      ],
      weeks.value.length,
    ),
    completionist: achievementLevelHelper(
      [
        [150, AchievementLevel.Gold],
        [80, AchievementLevel.Silver],
        [40, AchievementLevel.Bronze],
      ],
      uniqueVeggies.value.length,
    ),
    experimenterBean: achievementLevelHelper(
      [[15, AchievementLevel.Gold]],
      intersection(BEANS, uniqueVeggies.value).length,
    ),
    experimenterFruit: achievementLevelHelper(
      [[15, AchievementLevel.Gold]],
      intersection(FRUITS, uniqueVeggies.value).length,
    ),
    experimenterGrain: achievementLevelHelper(
      [[15, AchievementLevel.Gold]],
      intersection(GRAINS, uniqueVeggies.value).length,
    ),
    experimenterLeafy: achievementLevelHelper(
      [[15, AchievementLevel.Gold]],
      intersection(LEAFIES, uniqueVeggies.value).length,
    ),
    experimenterMushroom: achievementLevelHelper(
      [[15, AchievementLevel.Gold]],
      intersection(MUSHROOMS, uniqueVeggies.value).length,
    ),
    experimenterRoot: achievementLevelHelper(
      [[15, AchievementLevel.Gold]],
      intersection(ROOTS, uniqueVeggies.value).length,
    ),
    experimenterVegetable: achievementLevelHelper(
      [[15, AchievementLevel.Gold]],
      intersection(VEGETABLES, uniqueVeggies.value).length,
    ),
    hotStreak: achievementLevelHelper(
      [
        [20, AchievementLevel.Gold],
        [10, AchievementLevel.Silver],
        [5, AchievementLevel.Bronze],
      ],
      hotStreak.value,
    ),
    thousandsEven:
      allVeggies.value.length >= 1000 && Math.floor(allVeggies.value.length / 1000) % 2 === 0
        ? AchievementLevel.Platinum
        : AchievementLevel.NoAchievement,
    thousandsOdd:
      allVeggies.value.length >= 1000 && Math.floor(allVeggies.value.length / 1000) % 2 === 1
        ? AchievementLevel.Platinum
        : AchievementLevel.NoAchievement,
    ...weeklyAchievements.value(),
  }));

  // Actions
  const toggleVeggie = (targetVeggie: string) =>
    toggleVeggieForWeek(targetVeggie, currentWeekStart.value);

  const toggleVeggieForWeek = (targetVeggie: string, weekStart: DateTime) => {
    const targetWeek = weeks.value.find(({startDate}) => startDate.equals(weekStart));
    if (!targetWeek) {
      weeks.value = [
        ...weeks.value,
        {
          startDate: weekStart,
          veggies: [targetVeggie],
        },
      ];
      challenges.value = [
        ...challenges.value,
        {
          startDate: weekStart,
          veggie: getRandomVeggie(),
        },
      ];
    } else if (!targetWeek.veggies.includes(targetVeggie)) {
      targetWeek.veggies = [...targetWeek.veggies, targetVeggie];
    } else {
      targetWeek.veggies = difference(targetWeek.veggies, [targetVeggie]);
    }
  };

  const setVeggiesForWeek = (veggies: string[], weekStart: DateTime = currentWeekStart.value) => {
    const targetWeek = weeks.value.find(({startDate}) => startDate.equals(weekStart));
    if (!targetWeek) {
      weeks.value = [
        ...weeks.value,
        {
          startDate: weekStart,
          veggies,
        },
      ];
      challenges.value = [
        ...challenges.value,
        {
          startDate: weekStart,
          veggie: getRandomVeggie(),
        },
      ];
    } else {
      targetWeek.veggies = veggies;
    }
  };

  const $reset = () => {
    startDate.value = null;
    weeks.value = [];
    challenges.value = [];
  };

  return {
    achievements,
    allVeggies,
    atMostVeggies,
    challenges,
    completedChallenges,
    currentChallenge,
    currentVeggies,
    favorites,
    getWeekStarts,
    hotStreak,
    over30Veggies,
    startDate,
    suggestions,
    uniqueVeggies,
    veggiesForWeek,
    weeklyAchievements,
    weeks,
    setVeggiesForWeek,
    toggleVeggie,
    toggleVeggieForWeek,
    $reset,
  };
});
