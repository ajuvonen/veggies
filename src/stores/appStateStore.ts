import {ref, watchEffect} from 'vue';
import {defineStore, storeToRefs} from 'pinia';
import {useStorage} from '@vueuse/core';
import type {Settings} from '@/utils/types';
import {useAchievements} from '@/hooks/achievements';
import {useActivityStore} from '@/stores/activityStore';

type Message = {
  id: string;
  text: string;
};

export const useAppStateStore = defineStore('appState', () => {
  const {advance, achievements, reset} = useAchievements();
  const {uniqueVeggies} = storeToRefs(useActivityStore());

  watchEffect(() => {
    advance(uniqueVeggies.value, 0);
  });

  // State refs
  const settings = useStorage<Settings>(
    'veggies-settings',
    {
      locale: 'en',
    },
    localStorage,
    {
      mergeDefaults: true,
    },
  );

  const messages = ref<Message[]>([]);

  const addToastMessage = (text: string) => {
    messages.value.push({
      id: crypto.randomUUID(),
      text,
    });
  };

  const removeToastMessage = (targetId: string) => {
    messages.value = messages.value.filter(({id}) => id !== targetId);
  };

  const $reset = () => {
    messages.value = [];
    settings.value = {
      locale: 'en',
    };
    reset();
  };

  return {
    messages,
    settings,
    achievements,
    addToastMessage,
    removeToastMessage,
    $reset,
  };
});
