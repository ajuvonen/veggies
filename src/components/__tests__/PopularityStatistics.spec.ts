import {describe, it, expect, beforeEach} from 'vitest';
import {mount} from '@vue/test-utils';
import {DateTime} from 'luxon';
import {useActivityStore} from '@/stores/activityStore';
import PopularityStatistics from '@/components/PopularityStatistics.vue';
import {Category} from '@/utils/types';

const thisWeek = DateTime.now().startOf('week');
const lastWeek = DateTime.now().startOf('week').minus({weeks: 1});
const twoWeeksAgo = DateTime.now().startOf('week').minus({weeks: 2});

describe('PopularityStatistics', () => {
  let activityStore: ReturnType<typeof useActivityStore>;
  beforeEach(() => {
    activityStore = useActivityStore();
  });

  it('renders empty', () => {
    const wrapper = mount(PopularityStatistics);
    expect(wrapper.html()).toMatchSnapshot();
  });

  it('renders with data', () => {
    activityStore.startDate = thisWeek;
    activityStore.weeks.push({
      startDate: thisWeek,
      veggies: ['apple', 'cucumber', 'romaine', 'potato', 'red bean', 'teff'],
    });

    const wrapper = mount(PopularityStatistics);
    expect(wrapper.findByTestId('popularity-Fruit-1').find('.popularity__entry-text').text()).toBe(
      'apple (1)',
    );
    expect(
      wrapper.findByTestId('popularity-Vegetable-1').find('.popularity__entry-text').text(),
    ).toBe('cucumber (1)');
    expect(wrapper.findByTestId('popularity-Leafy-1').find('.popularity__entry-text').text()).toBe(
      'romaine (1)',
    );
    expect(wrapper.findByTestId('popularity-Root-1').find('.popularity__entry-text').text()).toBe(
      'potato (1)',
    );
    expect(wrapper.findByTestId('popularity-Bean-1').find('.popularity__entry-text').text()).toBe(
      'red bean (1)',
    );
    expect(wrapper.findByTestId('popularity-Grain-1').find('.popularity__entry-text').text()).toBe(
      'teff (1)',
    );

    Object.values(Category).forEach((category) => {
      expect(
        wrapper.findByTestId(`popularity-${category}-2`).find('.popularity__entry-text').text(),
      ).toBe('No Entry');
      expect(
        wrapper.findByTestId(`popularity-${category}-3`).find('.popularity__entry-text').text(),
      ).toBe('No Entry');
    });
  });

  it('renders entries in correct order', () => {
    activityStore.startDate = twoWeeksAgo;
    activityStore.weeks.push(
      {
        startDate: twoWeeksAgo,
        veggies: ['apple', 'lychee', 'pineapple'],
      },
      {
        startDate: lastWeek,
        veggies: ['apple', 'lychee', 'longan'],
      },
      {
        startDate: thisWeek,
        veggies: ['apple', 'blueberry', 'cloudberry'],
      },
    );

    const wrapper = mount(PopularityStatistics);
    expect(wrapper.findByTestId('popularity-Fruit-1').find('.popularity__entry-text').text()).toBe(
      'apple (3)',
    );
    expect(wrapper.findByTestId('popularity-Fruit-2').find('.popularity__entry-text').text()).toBe(
      'lychee (2)',
    );
    expect(wrapper.findByTestId('popularity-Fruit-3').find('.popularity__entry-text').text()).toBe(
      'pineapple (1)',
    );
  });
});
