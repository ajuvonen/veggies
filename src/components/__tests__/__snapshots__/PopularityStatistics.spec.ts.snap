// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`PopularityStatistics > renders empty 1`] = `
"<div data-v-4cf0b303="" class="popularity__container">
  <div data-v-4cf0b303="" class="flex-container flex-col">
    <h2 class="label-like" aria-hidden="false">Top-3 Fruits And Berries</h2>
    <ul data-v-4cf0b303="" class="flex-container flex-col align-start">
      <li data-v-4cf0b303="" class="popularity__entry" data-test-id="popularity-Fruit-1">
        <div data-v-4cf0b303="" class="popularity__entry-emoji" aria-hidden="true">ğŸ¥‡</div>
        <div data-v-4cf0b303="" class="popularity__entry-text">No Entry</div>
      </li>
      <li data-v-4cf0b303="" class="popularity__entry" data-test-id="popularity-Fruit-2">
        <div data-v-4cf0b303="" class="popularity__entry-emoji" aria-hidden="true">ğŸ¥ˆ</div>
        <div data-v-4cf0b303="" class="popularity__entry-text">No Entry</div>
      </li>
      <li data-v-4cf0b303="" class="popularity__entry" data-test-id="popularity-Fruit-3">
        <div data-v-4cf0b303="" class="popularity__entry-emoji" aria-hidden="true">ğŸ¥‰</div>
        <div data-v-4cf0b303="" class="popularity__entry-text">No Entry</div>
      </li>
    </ul>
  </div>
  <div data-v-4cf0b303="" class="flex-container flex-col">
    <h2 class="label-like" aria-hidden="false">Top-3 Vegetables</h2>
    <ul data-v-4cf0b303="" class="flex-container flex-col align-start">
      <li data-v-4cf0b303="" class="popularity__entry" data-test-id="popularity-Vegetable-1">
        <div data-v-4cf0b303="" class="popularity__entry-emoji" aria-hidden="true">ğŸ¥‡</div>
        <div data-v-4cf0b303="" class="popularity__entry-text">No Entry</div>
      </li>
      <li data-v-4cf0b303="" class="popularity__entry" data-test-id="popularity-Vegetable-2">
        <div data-v-4cf0b303="" class="popularity__entry-emoji" aria-hidden="true">ğŸ¥ˆ</div>
        <div data-v-4cf0b303="" class="popularity__entry-text">No Entry</div>
      </li>
      <li data-v-4cf0b303="" class="popularity__entry" data-test-id="popularity-Vegetable-3">
        <div data-v-4cf0b303="" class="popularity__entry-emoji" aria-hidden="true">ğŸ¥‰</div>
        <div data-v-4cf0b303="" class="popularity__entry-text">No Entry</div>
      </li>
    </ul>
  </div>
  <div data-v-4cf0b303="" class="flex-container flex-col">
    <h2 class="label-like" aria-hidden="false">Top-3 Leafy Greens And Herbs</h2>
    <ul data-v-4cf0b303="" class="flex-container flex-col align-start">
      <li data-v-4cf0b303="" class="popularity__entry" data-test-id="popularity-Leafy-1">
        <div data-v-4cf0b303="" class="popularity__entry-emoji" aria-hidden="true">ğŸ¥‡</div>
        <div data-v-4cf0b303="" class="popularity__entry-text">No Entry</div>
      </li>
      <li data-v-4cf0b303="" class="popularity__entry" data-test-id="popularity-Leafy-2">
        <div data-v-4cf0b303="" class="popularity__entry-emoji" aria-hidden="true">ğŸ¥ˆ</div>
        <div data-v-4cf0b303="" class="popularity__entry-text">No Entry</div>
      </li>
      <li data-v-4cf0b303="" class="popularity__entry" data-test-id="popularity-Leafy-3">
        <div data-v-4cf0b303="" class="popularity__entry-emoji" aria-hidden="true">ğŸ¥‰</div>
        <div data-v-4cf0b303="" class="popularity__entry-text">No Entry</div>
      </li>
    </ul>
  </div>
  <div data-v-4cf0b303="" class="flex-container flex-col">
    <h2 class="label-like" aria-hidden="false">Top-3 Roots And Bulbs</h2>
    <ul data-v-4cf0b303="" class="flex-container flex-col align-start">
      <li data-v-4cf0b303="" class="popularity__entry" data-test-id="popularity-Root-1">
        <div data-v-4cf0b303="" class="popularity__entry-emoji" aria-hidden="true">ğŸ¥‡</div>
        <div data-v-4cf0b303="" class="popularity__entry-text">No Entry</div>
      </li>
      <li data-v-4cf0b303="" class="popularity__entry" data-test-id="popularity-Root-2">
        <div data-v-4cf0b303="" class="popularity__entry-emoji" aria-hidden="true">ğŸ¥ˆ</div>
        <div data-v-4cf0b303="" class="popularity__entry-text">No Entry</div>
      </li>
      <li data-v-4cf0b303="" class="popularity__entry" data-test-id="popularity-Root-3">
        <div data-v-4cf0b303="" class="popularity__entry-emoji" aria-hidden="true">ğŸ¥‰</div>
        <div data-v-4cf0b303="" class="popularity__entry-text">No Entry</div>
      </li>
    </ul>
  </div>
  <div data-v-4cf0b303="" class="flex-container flex-col">
    <h2 class="label-like" aria-hidden="false">Top-3 Beans And Legumes</h2>
    <ul data-v-4cf0b303="" class="flex-container flex-col align-start">
      <li data-v-4cf0b303="" class="popularity__entry" data-test-id="popularity-Bean-1">
        <div data-v-4cf0b303="" class="popularity__entry-emoji" aria-hidden="true">ğŸ¥‡</div>
        <div data-v-4cf0b303="" class="popularity__entry-text">No Entry</div>
      </li>
      <li data-v-4cf0b303="" class="popularity__entry" data-test-id="popularity-Bean-2">
        <div data-v-4cf0b303="" class="popularity__entry-emoji" aria-hidden="true">ğŸ¥ˆ</div>
        <div data-v-4cf0b303="" class="popularity__entry-text">No Entry</div>
      </li>
      <li data-v-4cf0b303="" class="popularity__entry" data-test-id="popularity-Bean-3">
        <div data-v-4cf0b303="" class="popularity__entry-emoji" aria-hidden="true">ğŸ¥‰</div>
        <div data-v-4cf0b303="" class="popularity__entry-text">No Entry</div>
      </li>
    </ul>
  </div>
  <div data-v-4cf0b303="" class="flex-container flex-col">
    <h2 class="label-like" aria-hidden="false">Top-3 Grains, Nuts, And Seeds</h2>
    <ul data-v-4cf0b303="" class="flex-container flex-col align-start">
      <li data-v-4cf0b303="" class="popularity__entry" data-test-id="popularity-Grain-1">
        <div data-v-4cf0b303="" class="popularity__entry-emoji" aria-hidden="true">ğŸ¥‡</div>
        <div data-v-4cf0b303="" class="popularity__entry-text">No Entry</div>
      </li>
      <li data-v-4cf0b303="" class="popularity__entry" data-test-id="popularity-Grain-2">
        <div data-v-4cf0b303="" class="popularity__entry-emoji" aria-hidden="true">ğŸ¥ˆ</div>
        <div data-v-4cf0b303="" class="popularity__entry-text">No Entry</div>
      </li>
      <li data-v-4cf0b303="" class="popularity__entry" data-test-id="popularity-Grain-3">
        <div data-v-4cf0b303="" class="popularity__entry-emoji" aria-hidden="true">ğŸ¥‰</div>
        <div data-v-4cf0b303="" class="popularity__entry-text">No Entry</div>
      </li>
    </ul>
  </div>
</div>"
`;
